### ISSUE TITLE
CTS report retrieval fails with out of heap space

### ISSUE THREAD

**Issue by planetf1** [odpi/egeria#2497]
Running the CTS notebook within containers in kubernetes I hit:

```
http GET http://lab-dev:8080/servers/CTS_Server/open-metadata/conformance-suite/users/garygeeke/report
HTTP/1.1 500
Connection: close
Content-Type: application/json
Date: Fri, 24 Jan 2020 17:46:56 GMT
Transfer-Encoding: chunked

{
    "error": "Internal Server Error",
    "message": "Java heap space",
    "path": "/servers/CTS_Server/open-metadata/conformance-suite/users/garygeeke/report",
    "status": 500,
    "timestamp": "2020-01-24T17:46:56.111+0000"
}

```
```
lab-odpi-egeria-lab-dev-75f57db499-gxftc        5m           718Mi

kubectl describe pod/lab-odpi-egeria-lab-dev-75f57db499-gxftc | grep node
Node:         kube-node-57c2/172.31.3.224

kube-node-57c2     26m          2%     1664Mi          83%

```
Node is getting full ish but nothing is being reported
```
kubectl logs lab-odpi-egeria-lab-dev-75f57db499-gxftc | tail -2
Fri Jan 24 15:01:53 GMT 2020 CTS_Server Event OMRS-AUDIT-8006 Processing incoming event of type DeletedEntityEvent for instance 7ff7289d-6572-4681-9646-c5af63d84896 from: OMRSEventOriginator{metadataCollectionId='b80a4e5c-d25b-4261-9ad4-4b0faac49491', serverName='SUT_Server', serverType='Metadata Repository Server', organizationName='null'}
Fri Jan 24 15:01:53 GMT 2020 CTS_Server Event OMRS-AUDIT-8006 Processing incoming event of type PurgedEntityEvent for instance 7ff7289d-6572-4681-9646-c5af63d84896 from: OMRSEventOriginator{metadataCollectionId='b80a4e5c-d25b-4261-9ad4-4b0faac49491', serverName='SUT_Server', serverType='Metadata Repository Server', organizationName='null'}
```
And the server is still responding ok... suggesting we are hitting a lower limit than physical (if we were k8s would likely kill us / move us elsewhere)

**Comment by planetf1** [comment]
Note there is no exception logged, and by default debug logs are off, so it's tough to determine exact cause. I can try and repeat

**Comment by grahamwallis** [comment]
I'm not sure quite how to interpret the kubectl describe output - but it suggests that maybe you are running with 1.6G of heap - please confirm.

I don't have any heap issues when running on the JVM on a physical machine - and I have a very default configuration:
```
Grahams-MacBook-Pro:server grahamwallis$ java -Dserver.port=8082 -agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=5005   -jar server-chassis-spring-1.4-SNAPSHOT.jar
Grahams-MacBook-Pro:server grahamwallis$ env
MANPATH=/Users/grahamwallis/.nvm/versions/node/v11.4.0/share/man:/Library/Frameworks/Python.framework/Versions/2.7/share/man:/usr/local/share/man:/usr/share/man:/Library/Apple/usr/share/man:/Library/Developer/CommandLineTools/usr/share/man
TERM_PROGRAM=Apple_Terminal
NVM_CD_FLAGS=
TERM=xterm-256color
SHELL=/bin/bash
TMPDIR=/var/folders/ql/md81ttbd11n9rfmmynjlp7vc0000gn/T/
TERM_PROGRAM_VERSION=433
TERM_SESSION_ID=B944F544-24A0-4A72-A80B-0CC1D3E4A4E0
NVM_DIR=/Users/grahamwallis/.nvm
USER=grahamwallis
SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.XKG5tVfx88/Listeners
PATH=/Users/grahamwallis/.nvm/versions/node/v11.4.0/bin:/Library/Frameworks/Python.framework/Versions/2.7/bin:.:/Users/grahamwallis/.bundle:/Users/grahamwallis/.gem/ruby/2.6.0/bin:/Users/grahamwallis/.rbenv/shims:/Users/grahamwallis/.rbenv/bin:/Users/grahamwallis/dev/tools/apache-maven-3.5.4/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
LaunchInstanceID=73BE0F52-18CE-4DE6-BC98-1673687E8347
NVM_NODEJS_ORG_MIRROR=https://nodejs.org/dist
PWD=/Users/grahamwallis/dev/egeria-install/egeria-omag-1.4-SNAPSHOT/server
LANG=en_GB.UTF-8
XPC_FLAGS=0x0
XPC_SERVICE_NAME=0
HOME=/Users/grahamwallis
SHLVL=1
LOGNAME=grahamwallis
NVM_BIN=/Users/grahamwallis/.nvm/versions/node/v11.4.0/bin
NVM_IOJS_ORG_MIRROR=https://iojs.org/dist
SECURITYSESSIONID=186a9
OLDPWD=/Users/grahamwallis/dev/egeria-install
_=/usr/bin/env
Grahams-MacBook-Pro:server grahamwallis$ 
```

I don't find it surprising that there's no logging of an exception - exhaustion of heap space is fairly instant and terminal isn't it?

If it is possible to reproduce this consistently then it would be interesting to insert a breakpoint between the result collation and the return of the response object. As you suggest, pagination may help - but not the heap exhaustion occurs during result compilation. It would be good to know where it is occurring. 

But as a first step could you please confirm you heap size and help me understand why we see this in the containers but not in physical machines?

**Comment by planetf1** [comment]
I've only seen it ONCE in a container ... my suspicion is that perhaps we're getting closer to the limit... I've not tried again since friday.. We know the responses are pretty big so I can see potential for this being a lurking issue.

**Comment by grahamwallis** [comment]
Based on tests that Nigel and I ran yesterday I think we need to make a recommendation that the CTS Server has at least 1GB of heap.

**Comment by planetf1** [comment]
I'll assume I should update the docs. If you wish to please reassign ownership to yourself and remove me

**Comment by planetf1** [comment]
Added note in the CTS notebook as well as README

**Comment by planetf1** [comment]
This was fixed in #2651
